service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/delete/DeleteMetadataStartStep.java

```diff
@@ -0,0 +1,60 @@
+package bio.terra.workspace.service.resource.controlled.flight.delete;
+
+import static java.lang.Boolean.TRUE;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.db.ResourceDao;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A step for deleting the metadata that WSM stores for a controlled resource. */
+public class DeleteMetadataStartStep implements Step {
+
+  private final ResourceDao resourceDao;
+  private final UUID workspaceUuid;
+  private final UUID resourceId;
+
+  private final Logger logger = LoggerFactory.getLogger(DeleteMetadataStartStep.class);
+
+  public DeleteMetadataStartStep(ResourceDao resourceDao, UUID workspaceUuid, UUID resourceId) {
+    this.resourceDao = resourceDao;
+    this.workspaceUuid = workspaceUuid;
+    this.resourceId = resourceId;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    // State transition to deleting; or failure if another flight is using the resource
+    resourceDao.deleteResourceStart(workspaceUuid, resourceId, flightContext.getFlightId());
+    flightContext
+        .getWorkingMap()
+        .put(WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_STATE_CHANGED, TRUE);
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    // If we successfully changed state, then we assume that something bad happened
+    // during delete processing and call this a broken delete.
+    // Otherwise, we assume it failed because the resource is busy and simply return.
+    var resourceStateChanged =
+        flightContext
+            .getWorkingMap()
+            .get(WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_STATE_CHANGED, Boolean.class);
+    if (TRUE.equals(resourceStateChanged)) {
+      resourceDao.deleteResourceFailure(
```

Comment:
Right. I think there are problems with a lot of the undo. And in the current code, I don't know of any cases that will execute that code. The "surfacing the failure" code is a mistake and will immediately cause dismal failure even when one is not merited. Stairway already holds on to the original failure. Undo code is suppose to return success if it succeeds in performing the undo.

Reply:
I think we chose the "surface the failure" approach to avoid hiding the real failure when we have steps we know we can't undo, and a dismal failure is the right approach.
e.g. for `DeleteMetadataStep` we can't undo the delete, so returning `success` isn't right, but if the step throws a new exception then the flight's exception is overwritten with something unhelpful like "can't undo DeleteMetadataStep" instead of the real failure cause.  Does Stairway support separate do and undo exceptions?

Reply:
I believe it is not the case that returning a new FATAL err overwrites the real failure. If that is the behavior, it is a bug in Stairway. The intent is to preserve the original failure and return that as the flight status failure.

Reply:
Damn! I write a Stairway test and it does what you said it would. I see the Stairway bug and will fix with a Stairway ticket.

Reply:
I won't change this code, though.

Reply:
sgtm! I thought the overwriting was intended behavior in stairway, not using the "surface failure" pattern sounds right given that's actually a stairway bug